Add support for externally_connectable

kiaraarose opened the pull request:
#### b69952d3e355d35df99864dc9b7dbd33dd323f60
<pre>
Add support for externally_connectable
<a href="https://bugs.webkit.org/show_bug.cgi?id=268856">https://bugs.webkit.org/show_bug.cgi?id=268856</a>
<a href="https://rdar.apple.com/122422972">rdar://122422972</a>

Authors: Kiara Rose, Timothy Hatcher

Reviewed by Timothy Hatcher.

This patch creates seperate runtime and namespace objects to use for web pages
that wish to send and receive messages from extensions.

Also change how WebExtensionAPIObject tracks the world, and simplify child object creation
by passing the parent object. This touched all the API files, those changed files are elided.

Also change WebExtensionController to have a concept of testMode, so WebPage APIs can
use the browser.test API too, without an extensionContext. This required changing the
UI process WebExtensionContextAPITestCocoa to be WebExtensionControllerAPITestCocoa and
changing the delegates to no longer get a context. This wasn&apos;t used, so it is fine to remove.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:

* Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h:
(WebKit::toDebugString):
* Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.serialization.in:
Add a new WebExtensionContentWorldType::WebPage to distinguish messages being sent to and from web pages.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPortCocoa.mm:
(WebKit::WebExtensionContext::portPostMessage):
(WebKit::WebExtensionContext::firePortDisconnectEventIfNeeded):
Notify ports created from webPageRuntime.connect that the port was disconnected.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
(WebKit::WebExtensionContext::runtimeWebPageSendMessage):
(WebKit::WebExtensionContext::runtimeWebPageConnect):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::extensionContext const):
Mehtod to return extension context with a given unique identifier.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::WebExtensionMatchPattern::matchesPatternSetContainsMatchForURL):
Return true if one of the match pattern in the MatchPatternSet matches the given URL.

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIWebPageRuntime::sendMessage):
(WebKit::WebExtensionAPIWebPageRuntime::connect):

(WebKit::WebExtensionContextProxy::internalDispatchRuntimeMessageEvent):
Use the contentWorldType from the sendersParameters to determine which onConnect listeners
we should call.

(WebKit::WebExtensionContextProxy::dispatchRuntimeMessageEvent):
Event listeners cannot be added for the webPageRuntime namespace object.

(WebKit::WebExtensionContextProxy::internalDispatchRuntimeConnectEvent):
Use the contentWorldType from the sendersParameters to determine which onConnect listeners
we should call.

(WebKit::WebExtensionContextProxy::dispatchRuntimeConnectEvent):
Event listeners cannot be added for the webPageRuntime namespace object.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebPageNamespaceCocoa.mm:
Copied from Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h.
(WebKit::WebExtensionAPIWebPageNamespace::runtime):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebPageRuntimeCocoa.mm: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIObject.h:
(WebKit::WebExtensionAPIObject::WebExtensionAPIObject):

* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWebPageNamespace.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h.

* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWebPageRuntime.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h.

* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::toDOMWorld):

* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
(WebKit::WebExtensionControllerProxy::globalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::serviceWorkerGlobalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::addBindingsToWebPageFrameIfNecessary):
Inject custom browser namespace object for web pages.

* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWebPageNamespace.idl:
Copied from Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWebPageRuntime.idl:
Copied from Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h.
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.h:

Canonical link: <a href="https://commits.webkit.org/274898@main">https://commits.webkit.org/274898@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c2de9d4055b1bc676047b1b1179e81e3c81861b9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40321 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19333 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/42699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/42866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/36410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42628 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/22269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16662 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/42866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40895 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/22269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/42699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/42866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/22269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/42699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/44144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/22269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/42699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/44144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/12395 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/44144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/16755 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/42699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5338 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -19581,7 +19613,9 @@
 		5C400E6A29DB8AB500446F6F /* PBXTargetDependency */ = {
 			isa = PBXTargetDependency;
-			platformFilter = ios;
+			platformFilters = (
```
not sure why these changed. will revert

xeenon replied:
Xcode has been doing this to me too. Different versions fight over this stuff sometimes.

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -553,18 +670,23 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
-        auto& onConnect = namespaceObject.runtime().onConnect();
-        if (onConnect.listeners().isEmpty())
+        WebExtensionAPIEvent::ListenerVector listeners;
+        if (contentWorldType == WebExtensionContentWorldType::WebPage)
```
`sourceContentWorldType`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -439,6 +439,16 @@
+RefPtr<WebExtensionContext> WebExtensionController::extensionContext(const UniqueIdentifier& uniqueIdentifier) const
+{
+    for (Ref context : m_extensionContexts) {
+        if (context->isLoaded() && context->uniqueIdentifier() == uniqueIdentifier)
```
If it is in `m_extensionContexts` it is going to be loaded. So you can drop `context->isLoaded() && `.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
```diff
@@ -79,6 +79,8 @@ class WebExtensionMatchPattern : public API::ObjectImpl<API::Object::Type::WebEx
     static URLSchemeSet& validSchemes();
     static URLSchemeSet& supportedSchemes();
 
+    static bool matchPatternSetContainsMatchForURL(const MatchPatternSet&, URL&);
```
I'd call this (and move it up next to `static bool patternsMatchAllHosts()`:

```c++
static bool patternsMatchURL(const MatchPatternSet&, URL&);
```

Also `patternsMatchAllHosts` should use `const MatchPatternSet&` too!

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
```diff
@@ -244,6 +244,79 @@
+{
+    RefPtr destinationExtension = extensionController()->extensionContext(extensionID);
+    if (!destinationExtension) {
+        completionHandler(std::nullopt, @"extension not found");
```
Likely should return nothing here. `completionHandler(std::nullopt, std::nullopt);` we won't report it to the web page for privacy reasons.

kiaraarose replied:
Yup! I added this so I could tell that an issue occurred in the Web Process. I did mean to remove this call though
`callback->reportError(error.value());`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -396,6 +397,88 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
+        if (error) {
+            callback->reportError(error.value());
+            return;
+        }
```
We need to be careful about errors to the web page. I think just dropping this doing having `callback->call();` below happen on error is best.

kiaraarose replied:
meant to remove. will do

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -396,6 +397,88 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/runtime/connect
+
+    std::optional<String> name;
+    if (!WebExtensionAPIRuntime::parseConnectOptions(options, name, @"options", nil))
```
`outExceptionString` needs passed in here.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -574,16 +696,20 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
+
+    case WebExtensionContentWorldType::WebPage:
+        ASSERT_NOT_REACHED();
+        return;
```
You can group these:

```c++
case WebExtensionContentWorldType::Native:
case WebExtensionContentWorldType::WebPage:
    ASSERT_NOT_REACHED();
    return;
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
```diff
@@ -89,14 +90,35 @@ class WebExtensionAPIRuntime : public WebExtensionAPIObject, public WebExtension
+    RefPtr<WebExtensionAPIPort> connect(WebFrame&, JSContextRef, NSString *extensionID, NSDictionary *options, NSString **outExceptionString);
+
+private:
+    WebExtensionContextIdentifier m_contextIdentifier;
```
Not needed anymore.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWebPageRuntime.h:
```diff
@@ -0,0 +1,36 @@
+
+namespace WebKit {
+
+// FIXME: Remove once JSWebExtensionAPIWebPageNamespace.mm doesn't auto include it.
```
Oh, hmm… I'll fix this after you land.

xeenon commented on Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
```diff
@@ -141,6 +141,8 @@
         ASSERT_NOT_REACHED();
         return mainWorld();
+    case WebExtensionContentWorldType::WebPage:
+        return mainWorld();
```
Could move up and group with `case WebExtensionContentWorldType::Main`.

xeenon commented on Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
```diff
@@ -110,6 +112,21 @@
+    namespaceObject = toJS(context, WebExtensionAPIWebPageNamespace::create(WebExtensionAPIWebPageNamespace::ForMainWorld::Yes, WebExtensionContentWorldType::WebPage).ptr());
+
+    JSObjectSetProperty(context, globalObject, toJSString("browser").get(), namespaceObject, kJSPropertyAttributeNone, nullptr);
+    JSObjectSetProperty(context, globalObject, toJSString("chrome").get(), namespaceObject, kJSPropertyAttributeNone, nullptr);
```
We don't want to add `chrome` this will break a lot of pages and make them think Safari is Chrome.

xeenon commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWebPageNamespace.idl:
```diff
@@ -0,0 +1,33 @@
+
+    // Returns an object for the runtime object acccessibile to a webpage.
+    readonly attribute WebExtensionAPIWebPageRuntime runtime;
+};
```
Add a newline above.

xeenon commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWebPageRuntime.idl:
```diff
@@ -0,0 +1,33 @@
+
+    [RaisesException, NeedsFrame, ProcessArgumentsLeftToRight] void sendMessage(DOMString extensionId, [Serialization=JSON] any message, [Optional, NSDictionary] any options, [Optional, CallbackHandler] function callback);
+    [RaisesException, NeedsFrame, NeedsScriptContext] WebExtensionAPIPort connect(DOMString extensionId, [Optional, NSDictionary] any options);
+};
```
Add a newline above.

xeenon commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWebPageNamespace.idl:
```diff
@@ -0,0 +1,33 @@
+    ReturnsPromiseWhenCallbackIsOmitted,
+] interface WebExtensionAPIWebPageNamespace {
+
+    // Returns an object for the runtime object acccessibile to a webpage.
```
Comment not needed.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIRuntime.mm:
```diff
@@ -1136,6 +1154,69 @@
+        @"    browser.test.notifyPass()",
+        @"  })",
+        @"}, 1000)",
+        @"</script"
```
`</script>`

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIRuntime.mm:
```diff
@@ -1136,6 +1154,69 @@
+
+    [manager loadAndRun];
+
+    EXPECT_NS_EQUAL(manager.get().yieldMessage, @"Load Tab");
```
There is no `browser.test.yield('Load Tab')` in the background script.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
```diff
@@ -244,6 +244,79 @@
+
+    auto mainWorldProcesses = processes(WebExtensionEventListenerType::RuntimeOnConnectExternal, targetContentWorldType);
+    if (mainWorldProcesses.isEmpty()) {
+        completionHandler(toErrorString(@"runtime.connect()", nil, @"no runtime.onConnectExternal listeners found"));
```
@xeenon is this an error that we actually do want to report?

xeenon replied:
I don't think so. That should be just like the extension does not exist to the page.

kiaraarose commented on Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
```diff
@@ -52,23 +52,24 @@ class WebExtensionMatchPattern : public API::ObjectImpl<API::Object::Type::WebEx
 
-    static bool patternsMatchAllHosts(HashSet<Ref<WebExtensionMatchPattern>>&);
+    static bool patternsMatchAllHosts(const MatchPatternSet&);
+    static bool matchPatternSetContainsMatchForURL(const MatchPatternSet&, URL&);
```
still needs a name change `patternsMatchURL`

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
```diff
@@ -244,6 +244,87 @@
+    }
+
+    WebExtensionMessageSenderParameters completeSenderParameters = senderParameters;
+    if (RefPtr tab = getTab(senderParameters.pageProxyIdentifier))
```
i think an early return here if there is no tab is better. and update the permission check below to `!hasPermission(url, tab)`

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPortCocoa.mm:
```diff
@@ -39,6 +39,7 @@
 #import "WebExtensionAPINamespace.h"
 #import "WebExtensionAPIRuntime.h"
 #import "WebExtensionContextMessages.h"
+#import "WebExtensionControllerProxy.h"
```
remove

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -396,6 +397,89 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
+    // No options are supported currently.
+
+    WebExtensionMessageSenderParameters senderParameters {
+        std::nullopt, // unique identifer
```
identifier

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIRuntime.mm:
```diff
@@ -92,6 +92,24 @@
+    @"manifest_version": @3,
+
+    @"name": @"Externally Connectable Test",
+    @"description": @"Runtime Test",
```
"Externally Connectable Test"

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIRuntime.mm:
```diff
@@ -1136,6 +1154,72 @@
+        @"  browser.test.assertTrue(port?.sender?.origin?.startsWith('http'), 'sender.origin should start with http')",
+        @"  browser.test.assertEq(typeof port?.sender?.tab, 'object', 'sender.tab should be an object')",
+        @"  browser.test.assertEq(port?.sender?.frameId, 0, 'sender.frameId should be 0')",
+
```
remove extra newline

kiaraarose committed:
Add support for externally_connectable
https://bugs.webkit.org/show_bug.cgi?id=268856
rdar://122422972

Authors: Kiara Rose, Timothy Hatcher

Reviewed by Timothy Hatcher.

This patch creates seperate runtime and namespace objects to use for web pages
that wish to send and receive messages from extensions.

Also change how WebExtensionAPIObject tracks the world, and simplify child object creation
by passing the parent object. This touched all the API files, those changed files are elided.

Also change WebExtensionController to have a concept of testMode, so WebPage APIs can
use the browser.test API too, without an extensionContext. This required changing the
UI process WebExtensionContextAPITestCocoa to be WebExtensionControllerAPITestCocoa and
changing the delegates to no longer get a context. This wasn't used, so it is fine to remove.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:

* Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h:
(WebKit::toDebugString):
* Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.serialization.in:
Add a new WebExtensionContentWorldType::WebPage to distinguish messages being sent to and from web pages.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIPortCocoa.mm:
(WebKit::WebExtensionContext::portPostMessage):
(WebKit::WebExtensionContext::firePortDisconnectEventIfNeeded):
Notify ports created from webPageRuntime.connect that the port was disconnected.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm:
(WebKit::WebExtensionContext::runtimeWebPageSendMessage):
(WebKit::WebExtensionContext::runtimeWebPageConnect):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::extensionContext const):
Mehtod to return extension context with a given unique identifier.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMatchPatternCocoa.mm:
(WebKit::WebExtensionMatchPattern::matchesPatternSetContainsMatchForURL):
Return true if one of the match pattern in the MatchPatternSet matches the given URL.

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIWebPageRuntime::sendMessage):
(WebKit::WebExtensionAPIWebPageRuntime::connect):

(WebKit::WebExtensionContextProxy::internalDispatchRuntimeMessageEvent):
Use the contentWorldType from the sendersParameters to determine which onConnect listeners
we should call.

(WebKit::WebExtensionContextProxy::dispatchRuntimeMessageEvent):
Event listeners cannot be added for the webPageRuntime namespace object.

(WebKit::WebExtensionContextProxy::internalDispatchRuntimeConnectEvent):
Use the contentWorldType from the sendersParameters to determine which onConnect listeners
we should call.

(WebKit::WebExtensionContextProxy::dispatchRuntimeConnectEvent):
Event listeners cannot be added for the webPageRuntime namespace object.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebPageNamespaceCocoa.mm:
Copied from Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h.
(WebKit::WebExtensionAPIWebPageNamespace::runtime):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebPageRuntimeCocoa.mm: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIObject.h:
(WebKit::WebExtensionAPIObject::WebExtensionAPIObject):

* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWebPageNamespace.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h.

* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIWebPageRuntime.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h.

* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::toDOMWorld):

* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
(WebKit::WebExtensionControllerProxy::globalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::serviceWorkerGlobalObjectIsAvailableForFrame):
(WebKit::WebExtensionControllerProxy::addBindingsToWebPageFrameIfNecessary):
Inject custom browser namespace object for web pages.

* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWebPageNamespace.idl:
Copied from Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIWebPageRuntime.idl:
Copied from Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h.
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.h:

Canonical link: https://commits.webkit.org/274898@main
